server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -740,9 +740,9 @@ public enum SearcherScope {
     public abstract boolean isTranslogSyncNeeded();
 
     /**
-     * Ensures that all locations in the given stream have been written to the underlying storage.
+     * Ensures that the location has been written to the underlying storage.
      */
-    public abstract boolean ensureTranslogSynced(Stream<Translog.Location> locations) throws IOException;
+    public abstract void asyncEnsureTranslogSynced(Translog.Location location, Consumer<Exception> listener);
```

Comment:
Should we have a proper `ActionListener<Void>` instead of `Consumer<Exception>`?

Reply:
This would be a potential follow-up. But it would require many more changes as it is based on the `AsyncIOProcessor` so I would prefer not to include it in this commit.